lib/util/Strings.v3

```diff
@@ -43,6 +34,118 @@ component Strings {
 	def endsWithFrom(str: string, start: int, end: string) -> bool {
 		return (str.length - start) == end.length && endsWith(str, end);
 	}
+	// Splits a string on whitespace into an array of strings.
+	// maxsplit: Specifies how many splits to do. Pass -1 to split on "all occurrences."
+	def splitOnWhiteSpace(str: string, maxsplit: int) -> Array<string> {
+		if (str.length == 0) {
+			return [""];
+		}
+		if (maxsplit == 0) {
+			return [str];
+		}
+
+		maxsplit = if(maxsplit < 0, int.max, maxsplit);
+		var result = Vector<string>.new();
+		var timesMatched = 0;
+		var strStart = 0, strIdx = strStart;
+		while (strIdx < str.length) {
+			if (timesMatched >= maxsplit) {
+				// Finished matching based on max number of splits
+				break;
+			}
+			if (isWhiteSpace(str[strIdx])) {
+				timesMatched++;
+				if (strStart != strIdx) {
+					// matched whitespace, save split (if substring isn't empty string)
+					var substr = Arrays.range(str, strStart, strIdx);
+
+					result.put(substr);
+				}
+				str = Ranges.dup(Ranges.skipLeading(Arrays.range(str, strIdx, str.length), isWhiteSpace));
+				strStart = 0;
+				strIdx = strStart;
+			} else {
+				// did not match, continue looking for match
+				strIdx++;
+			}
+		}
+		if (strStart != strIdx || strIdx < str.length) {
+			// add rest of str to result
+			// don't add an empty string though if separating on WS!
+			var substr = Arrays.range(str, strStart, str.length);
+			result.put(substr);
+		}
+
+		return result.extract();
+	}
+	// Splits a string into an array of strings.
+	// charPred: the character predicate that determines which characters to split on in the string.
+	// maxsplit: Specifies how many splits to do. Pass -1 to split on "all occurrences."
+	def splitOnChar(str: string, charPred: (byte) -> bool, maxsplit: int) -> Array<string> {
+		return split_internal(str, maxsplit, split_char_cond(str, charPred, _), 1);
+	}
+	private def split_char_cond(str: string, charPred: (byte) -> bool, idx: int) -> bool {
+		return charPred(str[idx]);
+	}
+	// Splits a string into an array of strings.
+	// separator: Specifies the separator to use when splitting the string. Pass empty string to split on whitespace.
+	// maxsplit: Specifies how many splits to do. Pass -1 to split on "all occurrences."
+	def split(str: string, separator: string, maxsplit: int) -> Array<string> {
+		if (separator.length == 0) {
+			// split on whitespace
+			return splitOnWhiteSpace(str, maxsplit);
+		}
+		if (separator.length == 1) {
+			// split on single character
+			return splitOnChar(str, byte.==(separator[0], _), maxsplit);
+		}
+
+		return split_internal(str, maxsplit, split_cond(str, separator, _), separator.length);
+	}
+	private def split_cond(str: string, separator: string, idx: int) -> bool {
+		return startsWith(Arrays.range(str, idx, str.length), separator);
```

Comment:
That looks like it makes a copy of the string from the current index until the end, so that'll be expensive. Maybe just inline a byte-by-byte comparison here to avoid it?

Reply:
Yep, sgtm.